### PR TITLE
opt(rollup): change the way rollups are done

### PR DESCRIPTION
**Pull request by jarifibrahim**
**ISSUE:**
In a load that has a lot of updations to a key, a lot of deltas accumulate over time. These keys are added to the batch for a rollup. Due to the lossy behavior of rollups, they get dropped while the dgraph is busy rolling up keys (that might have much smaller delta count).

**Fix:**
This PR changes the way rollups are done.
- We now add keys to rollupBatch just after pushing the deltas over to badger.
- We now have a priority inside the keys that we rollup. The keys for which the deltas have increased up to a certain limit (hard-coded 500), we add it to high priority `rollupKeyPool` so that it has high chances to get rolled up and does not get dropped due to lossy behavior of rollups.

**Results:**
With a workload where a set of values are frequently updated:
- On master, the p directory size roars up to 6-7GB within 7-8 hours, and the `deltaCount` increases to 7-9k for lots of keys.
- On this branch, p directory size stays bounded in range 1-2GB even after >24 hours of workload.

---
This change is [Reviewable](https://reviewable.io/reviews/dgraph-io/dgraph/7253)

### CODE REVIEW

File: posting/mvcc.go [hunk 1]
```diff
@@ -328,7 +345,12 @@ func ReadPostingList(key []byte, it *badger.Iterator) (*List, error) {
 	deltaCount := 0
 	defer func() {
 		if deltaCount > 0 {
-			IncrRollup.addKeyToBatch(key)
+			// If deltaCount is high, send it to hihg priority channel instead.
```
Comment (jarifibrahim) [inline posting/mvcc.go:348]: hihg => high.

**Review by jarifibrahim — COMMENTED** [review]
LGTM. @NamanJain8 is the disk size bounded after this change?

**Comment by jarifibrahim** [comment]
This PR needs a proper title and description.

File: posting/mvcc.go [hunk 2]
```diff
@@ -150,15 +161,13 @@ func (ir *incrRollupi) Process(closer *z.Closer) {
...
-		case batch = <-ir.keysCh:
-			doRollup()
+		case batch := <-ir.priorityKeys[0].keysCh:
+			doRollup(batch, 0)
+			// Probably we don't need a limiter here because we expect not to call this function too frequently.
+
+		case batch := <-ir.priorityKeys[1].keysCh:
+			doRollup(batch, 1)
+
 			// throttle to 1 batch = 64 rollups per 100 ms.
```
Comment (ahsanbarkati) [inline posting/mvcc.go:171]: Is `1ms` sufficient throttle? Is so, fix this comment.

Comment (NamanJain8) [reply posting/mvcc.go]: Fixed the comment.

**Review by manishrjain — APPROVED** [review]
:lgtm: Just some comments.

> Reviewed 1 of 3 files at r5, 1 of 1 files at r6.
> *[Reviewable](https://reviewable.io/reviews/dgraph-io/dgraph/7253#-:-MQmI2zH1XDGjvDRX6Sj:b7o09xv)* status: all files reviewed, 3 unresolved discussions (waiting on @ahsanbarkati, @jarifibrahim, @martinmr, and @vvbalaji-dgraph)

___
*[posting/mvcc.go, line 136 at r6](https://reviewable.io/reviews/dgraph-io/dgraph/7253#-MQmFZ_H1KSPDaSYIE-C:-MQmFZ_I9G8JQBs4fU-M:be0972d) ([raw file](https://github.com/dgraph-io/dgraph/blob/7786424fef04b977096be594928cc33ddf53f808/posting/mvcc.go#L136)):*
> ```Go
> 
> 	doRollup := func(batch *[][]byte, priority int) {
> 		currTs := time.Now().Unix()
> ```

forcerollup is good to have. Use it over the priority pool.

Maybe add a TODO about dealing with opRollup not running.

**Comment by NamanJain8** [comment]
> LGTM. @NamanJain8 is the disk size bounded after this change?

Yes, it is. Mentioned that in PR description.

**Pull request merged**